rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -0,0 +1,322 @@
+    "name": "AWS_INSTANCE_ID",
+    "labels": {
+      "env": "example"
+      "teleport.dev/discovered-node": "yes"
+    },
+  },
+  "spec": {
+    "public_addr": "ec2-54-194-252-215.us-west-1.compute.amazonaws.com",
+    "hostname": "awsxyz"
+  }
+}
+```
+
+IAM join tokens will have a `tags` field added to the accounts so so only EC2
+instances with matching `tags` can be added using the token. the tags matching will
+default to `"*":"*"` to retain compatibility with previously generated IAM tokens
+
+## Agent installation
+
+In order to install the Teleport agent on EC2 instances, Teleport will serve an
+install script at `/webapi/installer`. This will be editable as a `tctl`
+resource.
+
+Example resource script:
+```yaml
+kind: installer
+metadata:
+spec:
+  # shell script that will be downloaded an run by the EC2 node
+  script: |
+    #!/bin/sh
+    curl https://.../teleport-pubkey.asc ...
+    echo "deb [signed-by=... stable main" | tee ... > /dev/null
+    apt-get update
+    apt-get install teleport
+    teleport ssh configure --auth-agent=...  --join-method=iam --token-name=iam-
+  # Any resource in Teleport can automatically expire.
+  expires: 0001-01-01T00:00:00Z
+```
+
+If a installer resource is configured to expire it will return an empty file
+after expiration.
+
+Unless overridden by a user, a default teleport installer command will be
+generated that is appropriate for the current running version and operating
+system initially supporting DEB and RPM based distros that Teleport already
+provides packages for.
+
+The user must create a custom SSM Command document that will be used to execute
+the served command.
+
+Example SSM aws:runCommand document:
+```yaml
+# name: installTeleport
+---
+schemaVersion: '2.2'
+description: aws:runShellScript
+mainSteps:
+- action: aws:downloadContent
+  name: downloadContent
+  inputs:
+    sourceType: "HTTP"
+    destinationPath: "/tmp/installTeleport.sh"
+    sourceInfo:
+      url: "https://teleportcluster.xyz/webapi/installer"
+- action: aws:runShellScript
+  name: runShellScript
+  inputs:
+    timeoutSeconds: '300'
+    runCommand:
+      - /bin/sh /tmp/installTeleport.sh
+
+```
+
+In order to run the new SSM document the AWS user will need IAM permissions to run
+SSM commands[3] for example:
+
+```json
+{
+    "Statement": [
+        {
+            "Action": "ssm:SendCommand",
+            "Effect": "Allow",
+            "Resource": [
+                # Allow running commands on all us-west-2 instances
+                "arn:aws:ssm:us-west-2:*:instance/*",
+                 # Allows running the installTeleport docuemnt on the allowed instances
+                "arn:aws:ssm:us-east-2:aws-account-ID:document/installTeleport"
+            ]
+        }
+    ]
+}
+```
+
+On AWS, Amazon Linux and Ubuntu LTS (16.04, 18.04, 20.04) come with the SSM agent
+preinstalled[4].
+
+## teleport.yaml generation
+
+A new `teleport` subcommand will also be added -- `teleport ssh configure`
```

Comment:
In order to clean up the `install.sh` script by moving some of the configuration generation to the `teleport` command I've created a [simple pr](https://github.com/gravitational/teleport/pull/13032) that adds a `teleport node configure` command. I can rename it as `ssh` to support this RFD.